scala/scala.bzl

```diff
@@ -384,3 +424,190 @@ def scala_repositories():
+    # scrooge_library tar, we can just use the sources directly
+    #TODO this will not give us the relative imports, though... we need
+    # to set a root dir for that to work properly
+    arguments=[f.path for f in ctx.attr.scrooge_library.srcs],
+  )
+
+  #TODO need to dive into what is actually necessary where
+  scalaattr = struct(outputs = ctx.outputs.srcjar, #TODO is this good to have here?
+                     transitive_runtime_deps = [], #TODO what are scrooge's runtime deps?
+                     transitive_compile_exports = transitive_cjars, #TODO are all necessary?
+                     transitive_runtime_exports = []
+                     )
+  return struct(scala = scalaattr, srcjar = ctx.outputs.srcjar)
+
+  #TODO the struct that we return from this is where we can set the scala attributes it wants, like the transitive dependencies
+
+
+scrooge_library = rule(
+  implementation = _scrooge_library_impl,
+  attrs = {
+      "srcs": attr.label_list(allow_files=_thrift_filetype),
+      "deps": attr.label_list(allow_files=_scroogelib_tar_filetype),
+      "_transitive_archive_files": attr.label_list(),
+  } + _implicit_deps,
+  outputs={"libarchive": "lib%{name}.scroogelib.tar"},
+)
+
+scrooge_scala_srcjar = rule(
+    _gen_scrooge_srcjar_impl,
+    attrs={
+        "scrooge_library": attr.label(
+            mandatory=True, providers=['srcs', '_transitive_archive_files']),
+        "_scrooge_core": attr.label(
+            default=Label("@scrooge_core//jar"),
+            allow_files=True),
+        "_libthrift": attr.label(
+            default=Label("@libthrift//jar"),
+            allow_files=True),
+        "_finagle_thrift": attr.label(
+            default=Label("@finagle_thrift//jar"),
+            allow_files=True),
+        "_scrooge_generator": attr.label(
+            default=Label("@scrooge_generator//jar"),
+            allow_files=True),
+        "_scopt": attr.label(
+            default=Label("@scopt//jar"),
+            allow_files=True),
+        "_twitter_util": attr.label(
+            default=Label("@twitter_util//jar"),
+            allow_files=True),
+        "_util_logging": attr.label(
+            default=Label("@util_logging//jar"),
+            allow_files=True),
+        "_scala_parser_combinators": attr.label(
+            default=Label("@scala_parser_combinators//jar"),
+            allow_files=True),
+    } + _implicit_deps,
+    outputs={
+      "srcjar": "lib%{name}.srcjar",
+      "manifest": "lib%{name}_MANIFEST.MF",
+    },
+)
+
+def scrooge_scala_library(name, scrooge_library, deps=[], visibility=None):
+  scrooge_scala_srcjar(
+    name = name + '_srcjar',
+    scrooge_library = scrooge_library,
+    visibility = visibility,
+  )
+  scala_library(
+    name = name,
+    srcjars = [name + '_srcjar'],
+    #TODO do these need to be here, given the transitive compile piece above?
+    deps = deps + [
+      "@libthrift//jar",
+      "@scrooge_core//jar",
+      "@finagle_thrift//jar",
```

Comment:
not sure if we want finagle on everything. Probably not. Probably an option to the rule.